KAFKA-2831; Do not use ZKUtils in `ConsumerGroupCommand` if `new-consumer` is used

ijuma opened the pull request.

ijuma committed:
Do not use ZKUtils in `ConsumerGroupCommand` if `new-consumer` is used

ijuma commented:
I'm still testing this, but created the PR to get some early reviews.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -41,306 +40,341 @@ object ConsumerGroupCommand {
+      else if (opts.options.has(opts.deleteOpt)) {
+        consumerGroupService match {
+          case service: ZkConsumerGroupService => service.delete()
+          case _ => throw new IllegalStateException(s"delete is not supported for $consumerGroupService")
```
nit: "not supported" might be the wrong way to phrase this. The new consumer deletes group metadata automatically when the last member leaves.

ijuma replied:
Thanks. How would you phrase it? Also note that this code should never run because of:

``` scala
if (options.has(newConsumerOpt)) {
        CommandLineUtils.checkRequiredArgs(parser, options, bootstrapServerOpt)
        if (options.has(deleteOpt))
          CommandLineUtils.printUsageAndDie(parser, "Option %s does not work with %s".format(deleteOpt, newConsumerOpt))
      } 
```

hachikuji replied:
I see. I guess it would be nice if we had a way to print a message that told the user that there is no need to delete group metadata for the new consumer, but not a big deal if that would require a lot of restructuring.

ijuma replied:
I updated the message when new-consumer is used with delete.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -41,306 +40,341 @@ object ConsumerGroupCommand {
+      topicPartitions
+        .sortBy { case topicPartition => topicPartition.partition }
+        .foreach { topicPartition =>
+          describePartition(group, topicPartition.topic, topicPartition.partition, getPartitionOffset(topicPartition),
```
A little annoying that we end up constructing a new KafkaConsumer for every partition. Every time we create it, we have to incur the overhead of looking up the coordinator and fetching topic metadata.

ijuma replied:
I changed it to always use the same consumer.

ijuma committed:
Reuse consumer and accept consumer.config option

Also improve error message if delete is used with new-consumer.

ijuma committed:
Remove `consumer.config` for now

It may be better to reuse `config`.

ijuma committed:
Fix NPE when committed returns null

hachikuji commented:
LGTM

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -41,306 +40,337 @@ object ConsumerGroupCommand {
+      val deserializer = (new StringDeserializer).getClass.getName
+      val brokerUrl = opts.options.valueOf(opts.bootstrapServerOpt)
+      properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerUrl)
+      properties.put(ConsumerConfig.GROUP_ID_CONFIG, "ConsumerGroupCommand")
```
Hmm... shouldn't we be getting this from the options? Otherwise committed() will just return the offsets for "ConsumerGroupCommand".

ijuma replied:
Yeah, I think you're right. That code was already there, but seems buggy.

hachikuji replied:
It probably would have been ok before if we were just checking log end offsets, but commits are group-specific.

ijuma replied:
Pushed a fix.

ijuma committed:
Set GROUP_ID_CONFIG in consumer correctly based on received option

Bug spotted by Jason.

ijuma commented:
@SinghAsDev, if you have the time, a review would be appreciated.

SinghAsDev commented:
@ijuma LGTM!

junrao commented:
Thanks for the patch. Hit the following exception.

bin/kafka-consumer-groups.sh --list --bootstrap-server localhost:9092  --new-consumer
Exception in thread "main" java.lang.NullPointerException
    at java.util.Hashtable.put(Hashtable.java:514)
    at kafka.admin.ConsumerGroupCommand$KafkaConsumerGroupService.createNewConsumer(ConsumerGroupCommand.scala:354)
    at kafka.admin.ConsumerGroupCommand$KafkaConsumerGroupService.<init>(ConsumerGroupCommand.scala:310)
    at kafka.admin.ConsumerGroupCommand$.main(ConsumerGroupCommand.scala:54)
    at kafka.admin.ConsumerGroupCommand.main(ConsumerGroupCommand.scala)

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -41,306 +40,337 @@ object ConsumerGroupCommand {
+      val deserializer = (new StringDeserializer).getClass.getName
+      val brokerUrl = opts.options.valueOf(opts.bootstrapServerOpt)
+      properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerUrl)
+      properties.put(ConsumerConfig.GROUP_ID_CONFIG, opts.options.valueOf(opts.groupOpt))
```
We probably need to use the previous default groupId when we're listing groups since no groupId would be provided in the options. Even better would be to not create a KafkaConsumer instance at all in that case.

ijuma replied:
I changed the code to create `KafkaConsumer` lazily and it is no longer created during `list()`.

ijuma committed:
Create consumer lazily in `KafkaConsumerGroupService`

We need a group-id to create a consumer, but a consumer is not
needed for `list()`. This avoids a NPE while trying to get a non-existent
group-id during `list()` spotted by Jun.

ijuma commented:
Thanks for the heads-up @junrao. I pushed a fix based on Jason's suggestion not to create the consumer unless needed. I tested both `list` and `describe` for old and new consumer and they both appear to work.

junrao commented:
Thanks for the patch. A few comments below.
1.  The following command succeeds. It would be useful to error out with a message on the invalid option "--bootstrap-server"
   bin/kafka-consumer-groups.sh --zookeeper localhost:2181 --list --bootstrap-server localhost:9092
2. The following command succeeds. It would be useful to error out with a message on the invalid option "--zookeeper"
   bin/kafka-consumer-groups.sh --zookeeper localhost:2181 --list --bootstrap-server localhost:9092 --new-consumer
3. Got the following error when describing a non-existing group before the offset topic is created. We can probably improve the output.
   bin/kafka-consumer-groups.sh --zookeeper localhost:2181 --bootstrap-server localhost:9092 --new-consumer --describe --group gggg

GROUP, TOPIC, PARTITION, CURRENT OFFSET, LOG END OFFSET, LAG, OWNER
Error while executing consumer group command The group coordinator is not available.
org.apache.kafka.common.errors.GroupCoordinatorNotAvailableException: The group coordinator is not available.

ijuma committed:
Improve error-handling

ijuma commented:
@junrao Fixed 1 and 2. I improved 3 a little (the header is not printed in that case). There are a number of exceptions that can be thrown when `adminClient.describeConsumerGroup(group)` is called. Is the issue that we are not handling exceptions in general or that the message is misleading in this particular case?

junrao commented:
Thanks for the latest patch. LGTM
